src/effects.ts

```diff
@@ -12,16 +12,82 @@ export function action<Action>(action: Action): Effect<Action> {
   return [dispatch => dispatch(action)];
 }
 
-export function fromPromise<Action>(
-  promise: () => Promise<Action>
+export function fromPromise<Action, Value = unknown, Err = Error>(
+  promise: () => Promise<Value>,
+  ofSuccess: (value: Value) => Action,
+  ofError: (error: Err) => Action
+): Effect<Action>
+
+export function fromPromise<Action, Value = unknown, Err = Error>(
+  promise: () => Promise<Value>,
```

Comment:
Action is returned from ofSuccess and ofError handlers, and the result of promise either resolved or rejected is wrapped with one of these handlers.

Reply:
But this overload doesn't take in an ofSuccess?

Reply:
ofSuccess is not necessary all the time, sometimes it is enough just to run an effect without state changes - https://github.com/iyegoroff/react-native-image-filter-kit/blob/9a732324433ab89ce6778647436585e0b3ddd6cf/examples/ImageExtraction/src/App.tsx#L51-L54

Reply:
Ah I see. That makes sense.

Reply:
However I do question whether having a default ofSuccess that swallows results is perhaps optimizing for the wrong thing? It it at the very least different to the original `fromPromise` design which assumed that you'd always return an `Action` from a promise.

Reply:
Last point on this. How about introducing another effect creator that is focused on void returning side effects. `doPromise(promise, ofError)`

Reply:
> However I do question whether having a default ofSuccess that swallows results is perhaps optimizing for the wrong thing?

It is similar to fromFunction without ofSuccess - sometimes there is no need to dispatch an action after successful promise resolving.

> which means that people upgrading might be in for a small surprise

But there are two overloads and people that need to return Action from promise can just use `fromPromise(promise, ofSuccess, ofError)` version all the time.

> How about introducing another effect creator that is focused on void returning side effects. doPromise(promise: () => Promise<void>, ofError). That'd leave the spirit of fromPromise the same, and reduce potential errors introduced by this breaking change.

I'm not sure why something like this is needed - fromPromise can be used with a promise returning void without problems. When library version is 0.x.y every change is considered breaking - https://semver.org/#spec-item-4, so if someone is using v0.2.3 they will continue to use this version after 0.2.4 release, unless they explicitly change library version in their package.json

Reply:
The issue I have is that the two overloads won't guide/force existing users
to the right usage of the API.

The problem can be illustrated by first giving an example of original usage
and then talking through the refactoring process.

Initial code:

```typescript
const apiEffect = Effects.fromPromise(fetchUserFromApi(id));
```

Now the user upgrades to the new version. Their code no longer compiles.
They see that fromPromise has been updated. They see two overloads. The one
has one less argument, and just needs an extra error handler. Since the
original fromPromise dispatched the success action automatically, a
reasonable assumption to make is that they should just add the error
handler and their action will be dispatched.

Of course after testing they'll eventually learn otherwise, but the API
design did not evolve in a way that guided them to the correct code without
having to run the application.

Reply:
Renamed `fromPromise` & `fromFunction` with two params to `attemptFunction` & `attemptPromise` to reduce potential confusion from breaking change.

Reply:
Thank you!. Apologise for all the back and forth, but want to get this API right.